lib/Parse/Parser.cpp

```
@@ -427,22 +430,50 @@ std::unique_ptr<PragmaDirective> Parser::parsePragmaDirective() {
   return std::make_unique<PragmaDirective>(SourceRange(Begin, End));
 }
 
-ContractDecl::ContractKind Parser::parseContractKind() {
+std::pair<ContractDecl::ContractKind, bool> Parser::parseContractKind() {
+  ContractDecl::ContractKind kind;
+  bool abstract = false;
+
+  if (Tok.getKind() == tok::kw_abstract) {
+    abstract = true;
+    ConsumeAnyToken();
+  }
+
   switch (Tok.getKind()) {
   case tok::kw_interface:
-    ConsumeToken();
-    return ContractDecl::ContractKind::Interface;
+    kind = ContractDecl::ContractKind::Interface;
+    break;
   case tok::kw_contract:
-    ConsumeToken();
-    return ContractDecl::ContractKind::Contract;
+    kind = ContractDecl::ContractKind::Contract;
+    break;
   case tok::kw_library:
-    ConsumeToken();
-    return ContractDecl::ContractKind::Library;
+    kind = ContractDecl::ContractKind::Library;
+    break;
   default:
     Diag(diag::err_expected_contract_kind);
-    ConsumeAnyToken();
-    return ContractDecl::ContractKind::Contract;
+    return std::make_pair(ContractDecl::ContractKind::Contract, false);
   }
+
+  ConsumeAnyToken();
+  return std::make_pair(kind, abstract);
+}
+
+std::unique_ptr<InheritanceSpecifier> Parser::parseInheritanceSpecifier() {
+  if (!Tok.isAnyIdentifier()) {
+    Diag(diag::err_expected) << tok::identifier;
+    return nullptr;
+  }
+  llvm::StringRef BaseName = Tok.getIdentifierInfo()->getName();
+  ConsumeToken(); // identifier
+
+  std::vector<std::unique_ptr<Expr>> Arguments;
+  if (isTokenParen()) {
+    ConsumeParen(); // (
+    // TODO : Constructor args is unimplement
```

Please raise an error message here.